vpr/src/route/sink_sampling.h

```diff
@@ -0,0 +1,155 @@
+            return true;
+        if (x > rhs.x)
+            return false;
+        if (y < rhs.y)
+            return true;
+        if (y > rhs.y)
+            return false;
+        return isink < rhs.isink;
+    }
+};
+
+/* Namespace the following to keep these generic fn names available */
+namespace sink_sampling {
+
+/* TODO: Replace this convex hull fn with something better
+ * it was cobbled together quickly, probably not the most efficient code */
+
+/** Cross product of v0v1 and v0p */
+constexpr int det(const SinkPoint& p, const SinkPoint& v0, const SinkPoint& v1) {
+    return (v1.x - v0.x) * (p.y - v0.y) - (v1.y - v0.y) * (p.x - v0.x);
+}
+
+/** Which side of [v0, v1] has p? +1 is right, -1 is left */
+constexpr int side_of(const SinkPoint& p, const SinkPoint& v0, const SinkPoint& v1) {
+    return det(p, v0, v1) > 0 ? 1 : -1;
+}
+
+/** Perpendicular distance of p to v0v1 assuming |v0v1| = 1
+ * (it's not, so only use to compare when v0 and v1 is the same for different p's) */
+inline int dist(const SinkPoint& p, const SinkPoint& v0, const SinkPoint& v1) {
+    return abs(det(p, v0, v1));
+}
+
+/** Helper for quickhull() */
+inline void find_hull(std::set<SinkPoint>& out, const std::vector<SinkPoint>& points, const SinkPoint& v0, const SinkPoint& v1, int side) {
+    int max_dist = 0;
+    const SinkPoint* max_p = nullptr;
+    for (auto& point : points) {
+        if (side_of(point, v0, v1) != side) {
+            continue;
+        }
+        int h = dist(point, v0, v1);
+        if (h > max_dist) {
+            max_dist = h;
+            max_p = &point;
+        }
+    }
+    if (!max_p) /* no point */
+        return;
+    out.insert(*max_p);
+    find_hull(out, points, v0, *max_p, -1);
+    find_hull(out, points, *max_p, v1, -1);
+}
+
+/** Find convex hull. Doesn't work with <3 points.
+ * See https://en.wikipedia.org/wiki/Quickhull */
+inline std::vector<SinkPoint> quickhull(const std::vector<SinkPoint>& points) {
+    if (points.size() < 3)
+        return std::vector<SinkPoint>();
+
+    std::set<SinkPoint> out;
+
+    int min_x = std::numeric_limits<int>::max();
+    int max_x = std::numeric_limits<int>::min();
+    const SinkPoint *min_p = nullptr, *max_p = nullptr;
+    for (auto& point : points) {
+        if (point.x <= min_x) {
+            min_x = point.x;
+            min_p = &point;
+        }
+        if (point.x >= max_x) {
+            max_x = point.x;
+            max_p = &point;
+        }
+    }
+    out.insert(*min_p);
+    out.insert(*max_p);
+    find_hull(out, points, *min_p, *max_p, -1);
+    find_hull(out, points, *min_p, *max_p, 1);
+    return std::vector<SinkPoint>(out.begin(), out.end());
+}
+
+} // namespace sink_sampling
+
+/** Which side of the cutline is this RRNode on?
+ * Cutlines are always assumed to be at cutline_axis = (cutline_pos + 0.5).
+ * In the context of the parallel router, a RR node is considered to be inside a bounding
+ * box if its (xlow, ylow) point is inside it. */
```

Comment:
You may need to update this as you will convert the RR node pruning to consider the drive point location.